### PR TITLE
RWA-1235 Excluded Grant Type Test Coverage

### PR DESCRIPTION
**Pull request by serdaroguzhanakin**
### JIRA link (if applicable) ###
https://tools.hmcts.net/jira/browse/RWA-1235

### Change description ###

**Does this PR introduce a breaking change?** (check one with "x")

```
[ ] Yes
[X] No
```

### CODE REVIEW

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/GetTaskByIdControllerTest.java [hunk 1]
```diff
@@ -93,4 +93,83 @@ public void should_return_a_200_with_task_and_correct_properties() {
         common.cleanUpTask(taskId);
     }
 
+
+    @Test
+    public void should_return_403_when_user_grant_type_standard_and_excluded() {
```
Comment (mkayad) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/GetTaskByIdControllerTest.java:98]: could be renamed as should_return_403_when_user_has_grant_type_standard_and_excluded

Comment (mkayad) [reply src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/GetTaskByIdControllerTest.java:98]: I think one more FT challenged and excluded could be added

**Review by mkayad — APPROVED** [review]
LGTM

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/GetTaskByIdControllerTest.java [hunk 2]
```diff
@@ -93,4 +93,83 @@ public void should_return_a_200_with_task_and_correct_properties() {
         common.cleanUpTask(taskId);
     }
 
+
```
Comment (BenNzewi) [inline src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/GetTaskByIdControllerTest.java:96]: What about the other search endpoints?

1. SearchForCriteria
2. SearchForCompleteable

Should't we add the same tests?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskTest.java [hunk 3]
```diff
@@ -0,0 +1,436 @@
...
+        AccessControlResponse accessControlResponse = new AccessControlResponse(null, roleAssignments);
+        permissionsRequired.add(PermissionTypes.CANCEL);
+
+        final Optional<TaskResource> task = cftQueryService.getTask(taskId, accessControlResponse, permissionsRequired);
+        Assertions.assertThat(task.isEmpty()).isTrue();
+
+    }
+
+    @Test
+    void should_retrieve_a_task_when_grant_type_specific() {
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskTest.java:201]: For Specific grant type we only have access to case. Should we be able to retrieve the task?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskTest.java [hunk 4]
```diff
@@ -0,0 +1,436 @@
...
+
+        final Optional<TaskResource> task = cftQueryService.getTask(taskId, accessControlResponse, permissionsRequired);
+        Assertions.assertThat(task.isPresent()).isTrue();
+        Assertions.assertThat(task.get().getTaskId()).isEqualTo(taskId);
+        Assertions.assertThat(task.get().getCaseId()).isEqualTo(caseId);
+
+    }
+
+    @Test
+    void should_retrieve_a_task_when_grant_type_specific_with_excluded() {
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskTest.java:230]: For Specific grant type we only have access to case. Should we be able to retrieve the task?

File: src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/GetTaskByIdControllerTest.java (hunk 2 above)
Comment (serdaroguzhanakin) [reply src/functionalTest/java/uk/gov/hmcts/reform/wataskmanagementapi/watasks/controllers/GetTaskByIdControllerTest.java:96]: We have another ticket for search endpoint ```https://tools.hmcts.net/jira/browse/RWA-1376```

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskTest.java [hunk 5]
```diff
@@ -0,0 +1,436 @@
...
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+
+@ActiveProfiles("integration")
+@DataJpaTest
+@AutoConfigureTestDatabase(replace = AutoConfigureTestDatabase.Replace.NONE)
+@Testcontainers
+@Sql("/scripts/wa/get_task_data.sql")
+public class CftQueryServiceGetTaskTest extends RoleAssignmentHelper {
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskTest.java:39]: The IT is called GetTask but the are tests for other action end points. Shouldn't these live in their own separate IT class?

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceSearchForCompletableTasksTest.java [hunk 6]
```diff
@@ -0,0 +1,111 @@
...
+
+        AccessControlResponse accessControlResponse = new AccessControlResponse(null, roleAssignments);
+        final GetTasksCompletableResponse<Task> task = cftQueryService.searchForCompletableTasks(searchEventAndCase,
+            accessControlResponse, permissionsRequired);
+        Assertions.assertThat(task).isNotNull();
+        Assertions.assertThat(task.isTaskRequiredForEvent()).isTrue();
+        Assertions.assertThat(task.getTasks().get(0).getCaseId()).isEqualTo(caseId);
+
+    }
+
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceSearchForCompletableTasksTest.java:100]: what about:

1. standard/excluded
2. challenged
3. challenged/excluded
4. specific
5. specific/excluded

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceSearchForCompletableTasksTest.java [hunk 7]
```diff
@@ -0,0 +1,111 @@
...
+        List<Map<String, CamundaVariable>> dmnResult = new ArrayList<>();
+        Map<String, CamundaVariable> response = Map.of(
+            "completionMode", new CamundaVariable("Auto", "String"),
+            "taskType", new CamundaVariable("processApplication", "String")
+        );
+        dmnResult.add(response);
+        return dmnResult;
+    }
+
+}
```
Comment (BenNzewi) [inline src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceSearchForCompletableTasksTest.java:111]: we don't have tests for endpoints:

1. SearchForCriteria

**Review by BenNzewi — COMMENTED** [review]
integrationTest/controllers does not have the following endpoints covered for standard, challenged, specific & excluded:

1. SearchForCriteria
2. SearchForCompleteable
3. GetTaskById
4. Complete
5. Cancel
6. Unclaim

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskTest.java (hunk 3 above)
Comment (serdaroguzhanakin) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskTest.java:201]: Specific grant type may have available tasks according to HLD

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskTest.java (hunk 4 above)
Comment (serdaroguzhanakin) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskTest.java:230]: Specific grant type may have available tasks according to HLD

File: src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskTest.java (hunk 5 above)
Comment (serdaroguzhanakin) [reply src/integrationTest/java/uk/gov/hmcts/reform/wataskmanagementapi/cft/query/wa/CftQueryServiceGetTaskTest.java:39]: will add separate classes for all

**Review by mkayad — APPROVED** [review]
LGTM

**Pull request merged**